File: sdk/ruby/lib/velocity_sdk/retry.rb
# frozen_string_literal: true

module VelocitySdk
  # Configuration for retry behavior with exponential backoff.
  #
  # @example
  #   policy = RetryPolicy.new(
  #     max_attempts: 5,
  #     initial_interval_ms: 100,
  #     backoff_coefficient: 2.0,
  #     max_interval_ms: 10_000,
  #     jitter: true,
  #   )
  #
  #   result = retry_with_policy(policy) { fetch_remote_data }
  class RetryPolicy
    attr_reader :max_attempts, :initial_interval_ms, :backoff_coefficient,
                :max_interval_ms, :jitter, :retryable_exceptions

    # @param max_attempts [Integer] Maximum number of attempts (>= 1)
    # @param initial_interval_ms [Numeric] Initial backoff interval in ms
    # @param backoff_coefficient [Numeric] Backoff multiplier (>= 1.0)
    # @param max_interval_ms [Numeric] Maximum backoff interval in ms
    # @param jitter [Boolean] Whether to add random jitter
    # @param retryable_exceptions [Array<Class>] Exception classes to retry on
    def initialize(
      max_attempts: 3,
      initial_interval_ms: 100,
      backoff_coefficient: 2.0,
      max_interval_ms: 60_000,
      jitter: true,
      retryable_exceptions: []
    )
      @max_attempts = max_attempts
      @initial_interval_ms = initial_interval_ms
      @backoff_coefficient = backoff_coefficient
      @max_interval_ms = max_interval_ms
      @jitter = jitter
      @retryable_exceptions = retryable_exceptions
      validate!
    end

    # Create a default retry policy.
    def self.defaults
      new
    end

    # Validate the policy configuration.
    # @raise [ArgumentError] if configuration is invalid
    def validate!
      raise ArgumentError, 'max_attempts must be >= 1' if @max_attempts < 1
      raise ArgumentError, 'initial_interval_ms must be > 0' if @initial_interval_ms <= 0
      raise ArgumentError, 'backoff_coefficient must be >= 1.0' if @backoff_coefficient < 1.0
      raise ArgumentError, 'max_interval_ms must be >= initial_interval_ms' if @max_interval_ms < @initial_interval_ms
    end

    # Calculate backoff duration for a given attempt (0-based).
    # @param attempt [Integer] Zero-based attempt index
    # @return [Numeric] Backoff duration in milliseconds
    def calculate_backoff(attempt)
      interval = @initial_interval_ms * (@backoff_coefficient**attempt)
      interval = [interval, @max_interval_ms].min

      if @jitter
        interval = rand * interval
      end

      interval
    end

    # Check if an exception is retryable.
    # @param error [Exception] The exception to check
    # @return [Boolean]
    def retryable?(error)
      return true if @retryable_exceptions.empty?
      @retryable_exceptions.any? { |klass| error.is_a?(klass) }
    end
  end

  # Execute a block with retry logic and exponential backoff.
  #
  # @param policy [RetryPolicy] Retry configuration
  # @yield The operation to execute
  # @return [Object] Result of the block
  # @raise [Exception] The last exception if all retries fail
  def self.retry_with_policy(policy, &block)
    policy.validate!

    last_error = nil

    policy.max_attempts.times do |attempt|
      begin
        return block.call
      rescue => e
        last_error = e

        raise e unless policy.retryable?(e)

        if attempt < policy.max_attempts - 1
          backoff_ms = policy.calculate_backoff(attempt)
          sleep(backoff_ms / 1000.0)
        end
      end
    end

    raise last_error
  end

  # Execute a block with default retry options.
  #
  # @param max_attempts [Integer] Maximum number of attempts
  # @param initial_interval_ms [Numeric] Initial backoff interval in ms
  # @param backoff_coefficient [Numeric] Backoff multiplier
  # @param max_interval_ms [Numeric] Maximum backoff interval in ms
  # @param jitter [Boolean] Whether to add random jitter
  # @yield The operation to execute
  # @return [Object] Result of the block
  def self.retry_with_backoff(
    max_attempts: 3,
    initial_interval_ms: 100,
    backoff_coefficient: 2.0,
    max_interval_ms: 60_000,
    jitter: true,
    &block
  )
    policy = RetryPolicy.new(
      max_attempts: max_attempts,
      initial_interval_ms: initial_interval_ms,
      backoff_coefficient: backoff_coefficient,
      max_interval_ms: max_interval_ms,
      jitter: jitter,
    )
    retry_with_policy(policy, &block)
  end
end

File: sdk/python/velocity_sdk/retry.py
"""
VELOCITY-WorkFlow Python SDK - Retry utilities.

Provides exponential backoff with jitter for retrying failed operations.
"""

import random
import time
from typing import Callable, Type, Set, Any
from dataclasses import dataclass, field


@dataclass
class RetryPolicy:
    """Configuration for retry behavior."""

    max_attempts: int = 3
    initial_interval: float = 0.1
    backoff_coefficient: float = 2.0
    max_interval: float = 60.0
    jitter: bool = True
    retryable_exceptions: Set[Type[Exception]] = field(default_factory=lambda: {Exception})

    def validate(self) -> None:
        """Validate retry policy configuration."""
        if self.max_attempts < 1:
            raise ValueError("max_attempts must be >= 1")
        if self.initial_interval <= 0:
            raise ValueError("initial_interval must be > 0")
        if self.backoff_coefficient < 1.0:
            raise ValueError("backoff_coefficient must be >= 1.0")
        if self.max_interval < self.initial_interval:
            raise ValueError("max_interval must be >= initial_interval")


def calculate_backoff(
    attempt: int,
    initial_interval: float,
    backoff_coefficient: float,
    max_interval: float,
    jitter: bool,
) -> float:
    """Calculate backoff duration for a given attempt."""
    interval = initial_interval * (backoff_coefficient ** attempt)
    interval = min(interval, max_interval)

    if jitter:
        # Full jitter: random value between 0 and calculated interval
        interval = random.uniform(0, interval)

    return interval


def retry_with_policy(
    policy: RetryPolicy,
    func: Callable[..., Any],
    *args: Any,
    **kwargs: Any,
) -> Any:
    """
    Execute a function with retry logic.

    Args:
        policy: Retry configuration
        func: Function to execute
        *args: Positional arguments for the function
        **kwargs: Keyword arguments for the function

    Returns:
        Result of the function call

    Raises:
        The last exception if all retries fail
    """
    policy.validate()

    last_exception: Exception | None = None

    for attempt in range(policy.max_attempts):
        try:
            return func(*args, **kwargs)
        except tuple(policy.retryable_exceptions) as e:
            last_exception = e

            if attempt < policy.max_attempts - 1:
                backoff = calculate_backoff(
                    attempt,
                    policy.initial_interval,
                    policy.backoff_coefficient,
                    policy.max_interval,
                    policy.jitter,
                )
                time.sleep(backoff)

    raise last_exception  # type: ignore[misc]


class RetryableOperation:
    """Context manager for retryable operations with custom error handling."""

    def __init__(self, policy: RetryPolicy):
        self.policy = policy
        self.attempts: list[Exception] = []

    def __call__(self, func: Callable[..., Any]) -> Callable[..., Any]:
        """Decorator for retryable functions."""
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            self.attempts.clear()
            return retry_with_policy(self.policy, func, *args, **kwargs)
        return wrapper

    def get_attempts(self) -> list[Exception]:
        """Get list of exceptions from previous attempts."""
        return self.attempts.copy()
